flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java

```diff
@@ -270,10 +270,12 @@ private void updateLastSeenWatermark(long timestamp) {
 	}
 
 	private void updateNFA(NFA<IN> nfa) throws IOException {
-		if (nfa.isEmpty()) {
-			nfaOperatorState.clear();
-		} else {
-			nfaOperatorState.update(nfa);
+		if (nfa.isNFAChanged()) {
+			if (nfa.isEmpty()) {
+				nfaOperatorState.clear();
+			} else {
+				nfaOperatorState.update(nfa);
```

Comment:
This is only true for the `RocksDB` state backend. If you are using the `Heap` state backend, you do not (de)serialize at every access but you keep the same object. So the flag should be re-set.